### ISSUE TITLE
[regression] file permission error with `quarto preview`

### ISSUE THREAD

**Issue by jmbuhr** [quarto-dev/quarto-cli#188]
Commit https://github.com/quarto-dev/quarto-cli/commit/b055f78d7d6bf78f88428534b6637a8bfcd5d59c introduced a regression for me (first time using  `git bisect`, this is pretty cool!). I am serving a website project with `quarto preview` and when I change a file I get the following output / error:

```
ERROR: Error: Directory not empty (os error 39)
    at deno:core/01_core.js:106:46
    at unwrapOpResult (deno:core/01_core.js:126:13)
    at Object.opSync (deno:core/01_core.js:140:12)
    at Object.removeSync (deno:runtime/js/30_fs.js:148:10)
    at removeIfExists (file:///hits/fast/mbm/share/mbm-wiki/_dev/quarto-cli/src/core/path.ts:23:10)
    at file:///hits/fast/mbm/share/mbm-wiki/_dev/quarto-cli/src/command/render/freeze.ts:201:23
    at Array.map (<anonymous>)
    at pruneProjectFreezerDir (file:///hits/fast/mbm/share/mbm-wiki/_dev/quarto-cli/src/command/render/freeze.ts:201:9)
    at freezeLibDir (file:///hits/fast/mbm/share/mbm-wiki/_dev/quarto-cli/src/command/render/project.ts:253:13)
    at renderProject (file:///hits/fast/mbm/share/mbm-wiki/_dev/quarto-cli/src/command/render/project.ts:258:11)
```
Since I have the quarto repo in the project as well we get the full stack trace for the typescript code.

I noticed that after the error occurs there is a rogue `site_libs` folder in the root directory of the project that I assume should have been copied into the `_site` folder and then be delete but this step failed due to some file permissions? When I then cancel with `ctrl+c` and run the `preview` command again the changes propagate and the folder disappears.

**Comment by jmbuhr** [comment]
Related information: I am the guy that is running quarto on a network drive: https://github.com/quarto-dev/quarto-cli/issues/109

**Comment by jjallaire** [comment]
Excellent! `git bisect` is indeed the bomb :-)

Hopefully I can narrow this down quickly....

**Comment by jjallaire** [comment]
That commit had some fundamental changes in how preview works (to accommodate previewing PDFs) so not something we can easily back out. My speculation is that the new ordering of things leaves a file that needs to be removed somewhere that it didn't before.

Interestingly, the call that is failing `Deno.removeSync` is actually documented to be able to recursively remove non-empty directories. There must be something about the network drive preventing this?

My proposed fix for the time being it to tolerate this error condition (while printing a warning): https://github.com/quarto-dev/quarto-cli/commit/254c0e7ce483cc422e6e4951103b73b68090f20b

LMK how this works on your end...

**Comment by jmbuhr** [comment]
What might make this hard is that the same commit works when not on a shared filesystem. I see no obvious reason why https://doc.deno.land/builtin/stable#Deno.removeSync fails. It should have write permissions, right? Otherwise there would be not `_site` at all. And since I am the user that started `quarto`, is there a reason it should not be able to to things that I can do?

**Comment by jjallaire** [comment]
I don't see an obvious reason either, but network filesystem interfaces have all sorts of vagaries and bugs in them so it's quite possible there is a part of the stack we don't understand or control that is causing the problem.

**Comment by jmbuhr** [comment]
I can confirm that this does indeed work. And it gives more information about the files affected.

```
WARNING: Unable to remove file /hits/fast/mbm/share/mbm-wiki/.quarto/_freeze/site_libs/bootstrap
WARNING: Unable to remove file /hits/fast/mbm/share/mbm-wiki/.quarto/_freeze/site_libs/quarto-html
WARNING: Unable to remove file /hits/fast/mbm/share/mbm-wiki/.quarto/_freeze/site_libs/quarto-nav
WARNING: Unable to remove file /hits/fast/mbm/share/mbm-wiki/.quarto/_freeze/site_libs/quarto-search
```

**Comment by jmbuhr** [comment]
Thank you for the quick fix! :)

**Comment by jjallaire** [comment]
That may be. Happily it's not at all a big deal if those files aren't removed. I think we'll keep the warning though as it will help identify other instances of this in the wild (and hopefully lead to a solution)

**Comment by jmbuhr** [comment]
TLDR: `Deno.removeSync` struggles with remote / shared file systems

Hi there, sorry to add to the backlog while rstudio::conf is going on (enjoy the conference!), but I am back with the same bug or rather a very similar one.

I am not 100% sure when this was introduced because. Currently I am on the latest version `v1.0.36`. When I run `quarto preview`, the folder `site_libs` is temporarily created before being moved into the `_site` folder.
However, the deletion of the libs folder or it's subdirectories can sometimes fail while editing files and navigating the preview site:
`Directory not empty (os error 39), remove '<...>/labbook/site_libs/clipboard'`

And sometimes a file seems to be treated as a directory:
`Not a directory (os error 20), stat '/hits/fast/mbm/buhrjk/phd/labbook/_site/notes.html/'`

Sometimes, but not always I get a warning before this happens:
`WARNING: connection closed before message completed`

I looked into this with a locally built version:

```
Error: Directory not empty (os error 39), remove '/hits/fast/mbm/buhrjk/phd/labbook/site_libs/quarto-search'
    at Object.opSync (deno:core/01_core.js:172:12)
    at Object.removeSync (deno:runtime/js/30_fs.js:150:10)
    at removeIfExists (file:///hits/fast/mbm/buhrjk/clones/quarto-cli/src/core/path.ts:29:10)
    at renderProject (file:///hits/fast/mbm/buhrjk/clones/quarto-cli/src/command/render/project.ts:369:17)
```

Indeed, if I guard this line with a try-catch this error is avoided (and the print statement I put in to see when it failed fired a couple of times per avoided crash):
https://github.com/quarto-dev/quarto-cli/blob/763b0ed7f060b206ef9b3a247612126b5f2f1f2a/src/core/path.ts#L29

It then crashed instead at:
https://github.com/quarto-dev/quarto-cli/blob/cf48867bae4b96ea90e0d8ade157cc7478ba56c9/src/command/render/project.ts#L374

though this is conditional on
https://github.com/quarto-dev/quarto-cli/blob/cf48867bae4b96ea90e0d8ade157cc7478ba56c9/src/command/render/project.ts#L290-L291
So it can be avoided by setting e.g. `keep-md: true` in quarto's frontmatter.

Maybe this is because of bandwidth limitations, it often works on reloading the page and correctly removes the folder until the next error. Putting all calls to `Deno.removeSync` in try-catch blocks seems like a very crude solution, but I haven't found a more elegant on, yet.
